compiler/src/test/java/com/squareup/anvil/compiler/ClassReferenceTest.kt

```diff
@@ -0,0 +1,291 @@
+
+          when (psiRef.shortName) {
+            "SomeClass1" -> {
+              assertThat(psiRef.functions).hasSize(0)
+              assertThat(descriptorRef.functions).hasSize(3)
+              assertThat(descriptorRef.functions.map { it.name }).containsExactly(
+                "equals", "toString", "hashCode"
+              )
+            }
+            "SomeClass2" -> {
+              assertThat(psiRef.functions).hasSize(1)
+              assertThat(descriptorRef.functions).hasSize(4)
+
+              val psiFunction = psiRef.functions.single()
+              val descriptorFunction = descriptorRef.functions.single { it.name == "test" }
+
+              assertThat(psiFunction.name).isEqualTo("test")
+              assertThat(descriptorFunction.name).isEqualTo("test")
+
+              assertThat(psiFunction.parameters).hasSize(1)
+              assertThat(descriptorFunction.parameters).hasSize(1)
+
+              assertThat(psiFunction.returnType().fqName).isEqualTo(Unit::class.fqName)
+              assertThat(descriptorFunction.returnType().fqName).isEqualTo(Unit::class.fqName)
+            }
+            "SomeInterface" -> {
+              assertThat(psiRef.functions).hasSize(1)
+              assertThat(descriptorRef.functions).hasSize(4)
+
+              val psiFunction = psiRef.functions.single()
+              val descriptorFunction = descriptorRef.functions.single { it.name == "test" }
+
+              assertThat(psiFunction.name).isEqualTo("test")
+              assertThat(descriptorFunction.name).isEqualTo("test")
+
+              assertThat(psiFunction.isAbstract()).isTrue()
+              assertThat(descriptorFunction.isAbstract()).isTrue()
+
+              assertThat(psiFunction.isConstructor()).isFalse()
+              assertThat(descriptorFunction.isConstructor()).isFalse()
+
+              assertThat(psiFunction.visibility()).isEqualTo(Visibility.PUBLIC)
+              assertThat(descriptorFunction.visibility()).isEqualTo(Visibility.PUBLIC)
+            }
+            "Companion" -> {
+              assertThat(psiRef.enclosingClass()?.shortName).isEqualTo("SomeInterface")
+              assertThat(descriptorRef.enclosingClass()?.shortName).isEqualTo("SomeInterface")
+
+              assertThat(psiRef.functions).hasSize(1)
+              assertThat(descriptorRef.functions).hasSize(4)
+            }
+            else -> throw NotImplementedError()
+          }
+
+          null
+        }
+      )
+    ) {
+      assertThat(exitCode).isEqualTo(OK)
+    }
+  }
+
+  @Test fun `the return type of a function can be resolved`() {
+    compile(
+      """
+      package com.squareup.test
+
+      class SomeClass1 {
+        fun test(string: String) = Unit
+      }
+
+      interface GenericInterface1<T> {
+        fun hello(): T
+      }
+
+      interface GenericInterface2<S> : GenericInterface1<S>
+
+      class SomeClass2 : GenericInterface1<String> {
+        override fun hello(): String = "hello"
```

Comment:
Types from the Kotlin std library are actually trickier to resolve, because they require no import 😅  These tests only cover a very small percentage of features and aren't complete by any means.